sway/sway_text_node.c

```diff
@@ -0,0 +1,291 @@
+		goto err;
+	}
+
+	cairo_t *cairo = cairo_create(surface);
+	cairo_set_antialias(cairo, CAIRO_ANTIALIAS_BEST);
+	cairo_set_font_options(cairo, fo);
+	pango = pango_cairo_create_context(cairo);
+
+	cairo_set_source_rgba(cairo, background[0], background[1], background[2], background[3]);
+	cairo_rectangle(cairo, 0, 0, width, height);
+	cairo_fill(cairo);
+
+	cairo_set_source_rgba(cairo, color[0], color[1], color[2], color[3]);
+	cairo_move_to(cairo, 0, (config->font_baseline - buffer->props.baseline) * scale);
+
+	render_text(cairo, config->font_description, scale, buffer->props.pango_markup,
+		"%s", buffer->text);
+
+	cairo_surface_flush(surface);
+
+	struct cairo_buffer *cairo_buffer = calloc(1, sizeof(*cairo_buffer));
+	if (!cairo_buffer) {
+		goto err;
+	}
+
+	wlr_buffer_init(&cairo_buffer->base, &cairo_buffer_impl, width, height);
+	cairo_buffer->surface = surface;
+	cairo_buffer->cairo = cairo;
+
+	wlr_scene_buffer_set_buffer(buffer->buffer_node, &cairo_buffer->base);
+	wlr_buffer_drop(&cairo_buffer->base);
+	update_source_box(buffer);
+
+	pixman_region32_t opaque;
+	pixman_region32_init(&opaque);
+	if (background[3] == 1) {
+		pixman_region32_union_rect(&opaque, &opaque, 0, 0,
+			buffer->props.width, buffer->props.height);
+	}
+	wlr_scene_buffer_set_opaque_region(buffer->buffer_node, &opaque);
+	pixman_region32_fini(&opaque);
+
+err:
+	if (pango) g_object_unref(pango);
+	cairo_font_options_destroy(fo);
+}
+
+static void handle_outputs_update(struct wl_listener *listener, void *data) {
+	struct text_buffer *buffer = wl_container_of(listener, buffer, outputs_update);
+	struct wlr_scene_outputs_update_event *event = data;
+
+	float scale = 0;
+	enum wl_output_subpixel subpixel = WL_OUTPUT_SUBPIXEL_UNKNOWN;
+
+	for (size_t i = 0; i < event->size; i++) {
+		struct wlr_scene_output *output = event->active[i];
+		if (subpixel == WL_OUTPUT_SUBPIXEL_UNKNOWN) {
+			subpixel = output->output->subpixel;
+		} else if (subpixel != output->output->subpixel) {
+			subpixel = WL_OUTPUT_SUBPIXEL_NONE;
+		}
+
+		if (scale != 0 && scale != output->output->scale) {
+			// drop down to gray scale if we encounter outputs with different
+			// scales or else we will have chromatic aberations
+			subpixel = WL_OUTPUT_SUBPIXEL_NONE;
+		}
+
+		if (scale < output->output->scale) {
+			scale = output->output->scale;
+		}
+	}
+
+	buffer->visible = event->size > 0;
+
+	if (scale != buffer->scale || subpixel != buffer->subpixel) {
+		buffer->scale = scale;
+		buffer->subpixel = subpixel;
+		render_backing_buffer(buffer);
+	}
+}
+
+static void handle_destroy(struct wl_listener *listener, void *data) {
+	struct text_buffer *buffer = wl_container_of(listener, buffer, destroy);
+
+	wl_list_remove(&buffer->outputs_update.link);
+	wl_list_remove(&buffer->destroy.link);
+
+	free(buffer->text);
+	free(buffer);
+}
+
+static void text_calc_size(struct text_buffer *buffer) {
+	struct sway_text_node *props = &buffer->props;
+
+	cairo_t *c = cairo_create(NULL);
```

Comment:
Can this fail?

Reply:
Yes.

Reply:
Can we add a check here, as we do for all other allocations?